docs/data/base/components/field/UnstyledFieldAsync.tsx

```diff
@@ -0,0 +1,137 @@
+import * as React from 'react';
+import * as Field from '@base_ui/react/Field';
+import { styled } from '@mui/system';
+
+const cache = new Map<string, string | null>();
+
+function checkAvailability(name: string) {
+  const takenNames = ['admin', 'root', 'superuser'];
+  return new Promise<string | null>((resolve) => {
+    setTimeout(() => {
+      const result = takenNames.includes(name) ? 'Name taken' : null;
+      cache.set(name, result);
+      resolve(result);
+    }, 500);
+  });
+}
+
+export default function UnstyledFieldAsync() {
+  const [loading, setLoading] = React.useState(false);
+
+  return (
+    <div>
+      <h3>Handle availability checker</h3>
+      <FieldRoot
+        validateOnChange
+        validate={async (value) => {
```

Comment:
Could you extract it to a function? Having long code snippets in JSX isn't too readable.